### ISSUE TITLE
Proxy leaks memory on 1.9.5 with constant ADS re-connections when DestinationRule TLS files are not mounted

### ISSUE THREAD

**Issue by chadlwilson** [istio/istio#33181]
**Bug description**

In Istio `1.9.5` when there is a `ServiceEntry` visible via `Sidecar` to a proxy, but whose `DestinationRule` cannot be fulfilled we see persistent memory leaks and constant ADS reconnects.

- proxied app `my-app` in the `services` namespace is trying to talk to `external.service.com` with direct sidecar egress
- We use namespaces and `Sidecar` to control visibility of various `ServiceEntry`s to various workloads
- The intent is for `my-app` to just talk directly to `external-service.com` without proxy TLS origination
- **However** due to a clash/incorrect configuration a `DestinationRule` for `external.service.com` is visible to the workload whose rule implies need for mutual TLS
- The required certs/keys defined by the `DestinationRule` are not mounted on the workload, so it cannot be loaded/used

Instead we see a persistent memory leak as soon as we upgraded to `1.9.5`
![image](https://user-images.githubusercontent.com/29788154/120008162-43d3e480-c00d-11eb-9f2a-0ca8402fa4dc.png)

Including constantly logging like the below
```
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:51:43.922194Z   info    ads     ADS: "@" sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-101 terminated rpc error: code = Canceled desc = context canceled
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:51:43.922502Z   warning envoy config    StreamSecrets gRPC config stream closed: 2, failed to generate secret for file-cert:/etc/external-service-client-cert/external-service-client-cert.pem~/etc/external-service-client-key/external-service-client-key.pem: open /etc/external-service-client-cert/external-service-client-cert.pem: no such file or directory
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:51:51.609032Z   info    ads     ADS: new connection for node:sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-102
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:51:53.317271Z   info    ads     ADS: new connection for node:sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-103
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:15.033899Z   error   cache   resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem failed to generate secret for proxy from file: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:15.034027Z   info    ads     ADS: "@" sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-102 terminated rpc error: code = Canceled desc = context canceled
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:15.034226Z   warning envoy config    StreamSecrets gRPC config stream closed: 2, failed to generate secret for file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:16.741779Z   error   cache   resource:file-cert:/etc/external-service-client-cert/external-service-client-cert.pem~/etc/external-service-client-key/external-service-client-key.pem failed to generate secret for proxy from file: open /etc/external-service-client-cert/external-service-client-cert.pem: no such file or directory
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:16.741893Z   info    ads     ADS: "@" sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-103 terminated rpc error: code = Canceled desc = context canceled
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:16.742205Z   warning envoy config    StreamSecrets gRPC config stream closed: 2, failed to generate secret for file-cert:/etc/external-service-client-cert/external-service-client-cert.pem~/etc/external-service-client-key/external-service-client-key.pem: open /etc/external-service-client-cert/external-service-client-cert.pem: no such file or directory
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:27.974668Z   info    ads     ADS: new connection for node:sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-104
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:30.566754Z   info    ads     ADS: new connection for node:sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-105
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:51.399660Z   error   cache   resource:file-cert:/etc/external-service-client-cert/external-service-client-cert.pem~/etc/external-service-client-key/external-service-client-key.pem failed to generate secret for proxy from file: open /etc/external-service-clie
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:51.399795Z   info    ads     ADS: "@" sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-104 terminated rpc error: code = Canceled desc = context canceled
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:51.399962Z   warning envoy config    StreamSecrets gRPC config stream closed: 2, failed to generate secret for file-cert:/etc/external-service-client-cert/external-service-client-cert.pem~/etc/external-service-client-key/external-service-client-key.pem: open /etc/external-service-client-cert/external-service-client-cert.pem: no such file or directory
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:53.991879Z   error   cache   resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem failed to generate secret for proxy from file: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:53.991991Z   info    ads     ADS: "@" sidecar~172.23.25.173~my-app-7f5795c6df-zsn5x.services~services.svc.cluster.local-105 terminated rpc error: code = Canceled desc = context canceled
my-app-7f5795c6df-zsn5x istio-proxy 2021-05-28T12:52:53.992185Z   warning envoy config    StreamSecrets gRPC config stream closed: 2, failed to generate secret for file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
```

The proxy keeps leaking 
- until it gets `OOMKilled`
- OR we remove the `DestinationRule`
- OR we remove the `Sidecar` visiblity of the `ServiceEntry`

[X] Networking
[X] Performance and Scalability

N/A

**Expected behavior**

The proxy should reject the `DestinationRule` and ignore it, but neither leak memory not cause constant reconnects.

On `1.8.6` the `DestinationRule` just got rejected and the proxy went about its life without issue. It logged a lot of noise, but that was it.

Logs from `1.8.6`
```
2021-05-20T03:59:56.842517Z	warning	envoy config	StreamSecrets gRPC config stream closed: 2, open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
2021-05-20T03:59:56.842354Z	info	sds	resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem connection is terminated: rpc error: code = Canceled desc = context canceled
2021-05-20T03:59:56.842251Z	error	sds	resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem Close connection. Failed to get secret for proxy "sidecar~172.23.26.183~my-app-7c6d7478b5-7t78g.services~services.svc.cluster.local" from secret cache: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
2021-05-20T03:59:56.842216Z	error	cache	resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem failed to generate secret for proxy from file: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
```

**Steps to reproduce the bug**

Config looks something like this.

```yaml
apiVersion: networking.istio.io/v1beta1
kind: ServiceEntry
metadata:
  name: https-egress
  namespace: services-egress
spec:
  hosts:
    - external.service.com
  location: MESH_EXTERNAL
  ports:
    - name: https
      number: 443
      protocol: HTTPS
  resolution: DNS
---
apiVersion: networking.istio.io/v1beta1
kind: Sidecar
metadata:
  name: my-app-sidecar-egress
  namespace: services
spec:
  egress:
    - hosts:
        - services-egress/external.service.com
        - ./*
        - istio-system/*
  workloadSelector:
    labels:
      app.kubernetes.io/name: my-app
---
apiVersion: networking.istio.io/v1beta1
kind: DestinationRule
metadata:
  name: originate-tls-for-external-service
  namespace: services
spec:
  exportTo:
    - .
  host: external.service.com
  trafficPolicy:
    portLevelSettings:
      - port:
          number: 443
        tls:
          caCertificates: /etc/external-service-trusted-ca-certs/caCertificates.pem
          clientCertificate: /etc/external-service-client-cert/external-service-client-cert.pem
          mode: MUTUAL
          privateKey: /etc/external-service-client-key/external-service-client-key.pem
          sni: external.service.com
```

May be related to #30504, however that is very generic and seems to refer to multiple Istio versions.

**Version**

```
$ istioctl version --remote
client version: 1.9.5
control plane version: 1.9.5
data plane version: 1.9.5 (31 proxies)

$ kubectl version --short
Client Version: v1.17.17
Server Version: v1.16.15-eks-ad4801
```

**How was Istio installed?**

`istioctl install -f operator.yaml`

**Environment where the bug was observed (cloud vendor, OS, etc)**

EKS

**Comment by hzxuzhonghu** [comment]
@chadlwilson Is that envoy memory increased?

**Comment by chadlwilson** [comment]
@hzxuzhonghu yes, sorry. I redacted the pod/container names. That is the proxy container memory usage for 3 replicas of the same `my-app`.

Forgot to mention that so far the leak seems to be independent of traffic through the proxies, i.e consistent growth over time no matter if they are idle or not.

Also, the CPU usage of the proxies also goes up over time as well, along with the memory.

**Comment by bianpengyuan** [comment]
Looks like an issue with envoy. Would be great if you can gather the heap profile for at least 15 mins (the longer the better) https://github.com/istio/istio/wiki/Analyzing-Istio-Performance#profile for envoy troubleshooting.

**Comment by howardjohn** [comment]
I strongly suspect the leak part is https://github.com/istio/envoy/commit/724e3094cbb746b57f164fb0a7ad9a76fe008868 which is not in 1.9.5.

**Comment by bianpengyuan** [comment]
> I strongly suspect the leak part is [istio/envoy@724e309](https://github.com/istio/envoy/commit/724e3094cbb746b57f164fb0a7ad9a76fe008868) which is not in 1.9.5.

@howardjohn did we see consistent increase of proxy memory with that bug?

**Comment by howardjohn** [comment]
I never observed it, but the XDS request is growing at a steady rate, so it seems it would leak?

Perhaps there are two leaks though.

**Comment by chadlwilson** [comment]
@howardjohn @bianpengyuan I believe I have managed to replicate with a vanilla istio `1.9.5` installation (`--set profile=demo` + prom + grafana addons) and `bookinfo` install using

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: no-certs
spec:
  host: google.com
  trafficPolicy:
    tls:
      mode: MUTUAL
      clientCertificate: /etc/certs/myclientcert.pem
      privateKey: /etc/certs/client_private_key.pem
      caCertificates: /etc/certs/rootcacerts.pem
---
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: google-com
spec:
  hosts:
  - google.com
  location: MESH_EXTERNAL
  ports:
  - number: 80
    name: http
    protocol: HTTP
```

Not 100% sure I am doing it right, but here's what I see. When I delete the `DestinationRule`+`ServiceEntry` subsequent profiles see no obvious leak anymore.

I do see `Node::MergeFrom` in the chain when it appears to be leaking.

![image](https://user-images.githubusercontent.com/29788154/121800830-a3193200-cc66-11eb-9347-a7d4ab67785b.png)


[profile010.pdf](https://github.com/istio/istio/files/6643551/profile010.pdf)

Are the constant ADS reconnects expected in such a case (not withstanding a leak being unexpected)?

**Comment by bianpengyuan** [comment]
Yeah looks like istio/envoy@724e309. Could you try 1.10 see if it fixes the issue?

**Comment by chadlwilson** [comment]
Yeah, I can't see the same leak on `1.10.1`; so I guess at least the memory leak part would be addressed by a release of https://github.com/istio/envoy/pull/321 and is related to https://github.com/istio/istio/issues/31904

Now looks like:
![image](https://user-images.githubusercontent.com/29788154/121836965-24280600-cd07-11eb-8cb7-b4cb48958bf1.png)
[profile014.pdf](https://github.com/istio/istio/files/6645517/profile014.pdf)


So I guess my remaining possible concern is the ADS reconnects. On `1.10.1` we still see

```
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:20.398794Z	error	cache	resource:file-cert:/etc/certs/myclientcert.pem~/etc/certs/client_private_key.pem failed to generate secret for proxy from file: open /etc/certs/myclientcert.pem: no such file or directory
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:20.399079Z	info	ads	ADS: "@" sidecar~10.42.2.19~ratings-v1-6549d464dc-zc4v9.default~default.svc.cluster.local-73 terminated rpc error: code = Canceled desc = context canceled
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:20.399497Z	warning	envoy config	StreamSecrets gRPC config stream closed: 2, failed to generate secret for file-cert:/etc/certs/myclientcert.pem~/etc/certs/client_private_key.pem: open /etc/certs/myclientcert.pem: no such file or directory
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:32.972134Z	error	cache	resource:file-root:/etc/certs/rootcacerts.pem failed to generate secret for proxy from file: open /etc/certs/rootcacerts.pem: no such file or directory
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:32.972370Z	info	ads	ADS: "@" sidecar~10.42.2.19~ratings-v1-6549d464dc-zc4v9.default~default.svc.cluster.local-74 terminated rpc error: code = Canceled desc = context canceled
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:32.972861Z	warning	envoy config	StreamSecrets gRPC config stream closed: 2, failed to generate secret for file-root:/etc/certs/rootcacerts.pem: open /etc/certs/rootcacerts.pem: no such file or directory
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:33.071490Z	info	ads	ADS: new connection for node:sidecar~10.42.2.19~ratings-v1-6549d464dc-zc4v9.default~default.svc.cluster.local-75
ratings-v1-6549d464dc-zc4v9 istio-proxy 2021-06-14T03:50:43.365910Z	info	ads	ADS: new connection for node:sidecar~10.42.2.19~ratings-v1-6549d464dc-zc4v9.default~default.svc.cluster.local-76
```

On `1.8.6` by contrast it just kept logging stuff like
```
2021-05-20T03:59:56.842517Z	warning	envoy config	StreamSecrets gRPC config stream closed: 2, open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
2021-05-20T03:59:56.842354Z	info	sds	resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem connection is terminated: rpc error: code = Canceled desc = context canceled
2021-05-20T03:59:56.842251Z	error	sds	resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem Close connection. Failed to get secret for proxy "sidecar~172.23.26.183~my-app-7c6d7478b5-7t78g.services~services.svc.cluster.local" from secret cache: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
2021-05-20T03:59:56.842216Z	error	cache	resource:file-root:/etc/external-service-trusted-ca-certs/caCertificates.pem failed to generate secret for proxy from file: open /etc/external-service-trusted-ca-certs/caCertificates.pem: no such file or directory
```

Is this just logging noise, or is this unexpected behaviour by Envoy e.g the "invalid" `DestinationRule` config is causing it to invalidate the connections and reconnect?

**Comment by chadlwilson** [comment]
FWIW `1.10.1` is potentially a bit of an issue for us due to the [Kubernetes support](https://istio.io/latest/docs/releases/supported-releases/#support-status-of-istio-releases). Some of our clusters are still running `1.16` (don't ask - should be resolved soonish) so we already took on a bit of risk by upgrading to the _technically_ unsupported configuration of `(k8s 1.16, istio 1.9.x)` - we're unsure of whether `(k8s 1.16, istio 1.10.x)` might be pushing it a bit far out of the tested configurations.

**Comment by chadlwilson** [comment]
Hi @howardjohn @bianpengyuan  - would the underlying fix for this in `istio/envoy` now be in `1.9.6`? Not really sure based on the tags in `istio/envoy` and the [bump](https://github.com/istio/istio/pull/33063/files) to `1.9-dev.9`. :(

**Comment by howardjohn** [comment]
Yeah it should be in 1.9.6, that has envoy sha a5f131fc2e13c0ebdf77bbc064e503f248ff5d2 which has the PR

**Comment by chadlwilson** [comment]
Much appreciated @howardjohn - the leak does seem to be fixed with `1.9.6`. I will close this accordingly, as this was my main concern, although the constant ADS reconnects seem a little odd.

I'm not sure if those are expected in the case of a configuration that a proxy is rejecting. If unexpected I can raise a separate issue. There are also no errors reported by pilot metrics or `istioctl proxy-status`. I'm sure historically these types of errors resulted in a sync error of some sort, but perhaps that was intentional.

**Comment by chadlwilson** [comment]
Closing, as patch for memory leak is included in Istio `1.9.6` and `1.10.1`
